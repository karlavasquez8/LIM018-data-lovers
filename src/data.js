// datos de la tarjeta
export const createCard = (pokemon) => {
  const {img, name, num, type,weaknesses,resistant } = pokemon;
  return `
      <section class="info-card-pokemon"> 
      <div class="img-pokemon same-style">
        <img class="pokemon" src="${img}" loading="lazy">
      </div>
      <div class="data-pokemon">
        <span class="name-pokemon description-pokemon">${name}</span>
        <span class="description-pokemon">${num}</span>
        <span class="description-pokemon">Tipo: ${type.join(', ')} </span>
        <span class="description-pokemon">Resistencia: ${resistant.join(', ')} </span>
        <span class="description-pokemon">Debilidad: ${weaknesses.join(', ')} </span>
      </div>
    </section>
  `;
};

// ordenando alfabeticamente de la A a la Z.
export function sortByAz(pokemonAsc){
 const pokemonAz = pokemonAsc.sort((a, b)=>{
  if(a.name < b.name){
     return -1;
  }
  if (a.name > b.name){
    return 1;
  }
     return 0;
 }
 );

 const pokemonAzCards = pokemonAz.map((nombreAz)=>createCard(nombreAz));
 return pokemonAzCards.join('');
}
// ordenando de forma alfabetica de la Z a la A.

export function sortByZa(pokemonsDes){
 const pokemonZa = pokemonsDes.sort((a, b)=>{
      if(a.name > b.name){
          return -1;
       }
       if(a.name < b.name){
          return 1;
       }
          return 0;
    }
    );
  const pokemonZaCards = pokemonZa.map((nombreZa)=>createCard(nombreZa));
  return pokemonZaCards.join('');
}

export function sortBy(pokemons, order){
  const sortPokemons = pokemons.sort((a, b)=>{
    const conditionalOne = order == "desc"? a.name > b.name : a.name < b.name;
    const conditionalTwo = order == "desc"? a.name < b.name : a.name > b.name;
    if(conditionalOne){
      return -1;
    }
    if (conditionalTwo) {
      return 1;
    }
  return 0;
  }
 );

 return sortPokemons;
}

// botones del filtrado en modal
export const createButton = (filters, group)=>{
  return `
  <button class="filter-option font-color" data-group="${group}">${filters}</button>
  `;
}

//filtrar datos por tipos, generacion, resistencia, debilidades y huevos

  const filterType = (pokemons, type) => {
    return pokemons.filter((pokemon) => pokemon.type.includes(type))
  };

 //const filterGeneration= (pokemons, generation) => {
   // return pokemons.filter((pokemon) => pokemon.generation == generation)
  //};

  const filterResistant= (pokemons, resistant) => {
    return pokemons.filter((pokemon) => pokemon.resistant.includes(resistant))
  };

  const filterWeaknesses= (pokemons, weaknesses) => {
    return pokemons.filter((pokemon) => pokemon.weaknesses.includes(weaknesses))
  };

  //const filterEggs= (pokemons, eggs) => {
    //return pokemons.filter((pokemon) => pokemon.eggs == eggs)
  //};

  const getFiltersItems = (pokemons) =>{
    const items = {
      types:[],
      resistant:[],
      weaknesses:[],
    }
    return pokemons.reduce((prev,pokemon)=>{
      prev.types.push(...pokemon.type)
      prev.resistant.push(...pokemon.resistant)
      prev.weaknesses.push(...pokemon.weaknesses)

      return {
        types:Array.from(new Set(prev.types)),
        resistant:Array.from(new Set(prev.resistant)),
        weaknesses:Array.from(new Set(prev.weaknesses)),
      }
    },items)
  }
  
export const filters = {
  filterType,
  //filterGeneration,
  filterResistant,
  filterWeaknesses,
  //filterEggs
}
export const filtersItems = {
  getFiltersItems,
}

// porcentaje de pokemones

export function computeStats (numPokemons, numTypes){
   return parseFloat((numTypes * 100 / numPokemons).toFixed(2));
}
//esta funcion me retorna un numero pero con el toFixed le convierte en un string)

//funcion para el buscador

export function filtrarSearch(pokemons, name){
 return pokemons.filter(pokemon =>pokemon.name.startsWith(name));
}







