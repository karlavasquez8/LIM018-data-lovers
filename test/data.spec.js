import { describe, it } from 'eslint/lib/rule-tester/rule-tester';
import { createCard,createButton, computeStats, sortByAz} from '../src/data.js';
//import data from '../src/data/pokemon/pokemon.js';

describe('createCard', () => {
  it('should by a function', () => {
    expect(typeof createCard).toBe('function');
  });

  it('should return a pokemon card', () => {
    const pokemon = {
      "num": "001",
      "name": "bulbasaur",
      "generation": {
        "num": "generation i",
        "name": "kanto"
      },
      "about": "Bulbasaur can be seen napping in bright sunlight. There is a seed on its back. By soaking up the sun's rays, the seed grows progressively larger.",
      "img": "https://www.serebii.net/pokemongo/pokemon/001.png",
      "size": {
        "height": "0.71 m",
        "weight": "6.9 kg"},
      "type":[
          "grass",
          "poison"
      ],
      "resistant": [
        "water",
        "electric",
        "grass",
        "fighting",
        "fairy"
      ],
      "weaknesses": [
        "fire",
        "ice",
        "flying",
        "psychic"
      ]}
    expect(createCard(pokemon)).toMatchSnapshot()
  
  });
});
describe('sortByAz',() =>{
  it('should by a function',() =>{
    expect(typeof sortByAz).toBe('function');
  });

  it('')
})

describe('createButton', () => {
  it('should by a function', () => {
    expect(typeof createButton).toBe('function');
  });

  it('should returns name Pokemon and group', () => {
    expect(createButton("grass","type")).toContain("grass");
    expect(createButton("grass","type")).toContain("type");
  });
});

describe('computeStats',() => {
  it('should by a function',() => {
    expect (typeof computeStats).toBe('function');
  });

  it('should return a calculation',()=>{
    expect(computeStats(251,17)).toBe(6.77)
  })
})
