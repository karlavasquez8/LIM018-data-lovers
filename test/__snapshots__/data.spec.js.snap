// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCard should return a pokemon card 1`] = `
"
      <section class=\\"info-card-pokemon\\"> 
      <div class=\\"img-pokemon same-style\\">
        <img class=\\"pokemon\\" src=\\"https://www.serebii.net/pokemongo/pokemon/001.png\\" loading=\\"lazy\\">
      </div>
      <div class=\\"data-pokemon\\">
        <span class=\\"name-pokemon description-pokemon\\">bulbasaur</span>
        <span class=\\"description-pokemon\\">001</span>
        <span class=\\"description-pokemon\\">Tipo: grass, poison </span>
        <span class=\\"description-pokemon\\">Resistencia: water, electric, grass, fighting, fairy </span>
        <span class=\\"description-pokemon\\">Debilidad: fire, ice, flying, psychic </span>
      </div>
    </section>
  "
`;
